### PR TITLE
[muon] add multi-tensor (foreach) standard Muon path (#187131)

### PR DESCRIPTION
**Pull request by tangyuhao**
Summary:

Adds a foreach implementation for `torch.optim.Muon` and uses PyTorch standard automatic foreach selection when `foreach=None` on supported devices. Users can still pass `foreach=False` to force the single-tensor path.

This change introduces `_multi_tensor_muon` and wires `foreach` through `Muon`, the functional API, and state-dict BC. The foreach path fuses momentum, weight decay, and shape-grouped parameter updates while preserving per-tensor Newton-Schulz computation and numerical parity with the single-tensor path.

It enables generic foreach parity and BC coverage for Muon, adapts generic optimizer tests to valid 2D Muon parameters, extends the generic large-tensor stress test to cover more than `2**32` Muon elements without allowing a silent no-op to pass, and adds CUDA microbenchmarks comparing foreach with the for-loop implementation. The `fbcode/caffe2/...` and `xplat/caffe2/...` mirrors remain synchronized.

Test Plan:
CPU optimizer tests:
`buck2 --isolation-dir codex-muon-bench test fbcode//mode/opt fbcode//caffe2/test:optim -- --regex Muon`
Result: 43 passed, 0 failed, 8 skipped.
TestInfra: https://www.internalfb.com/intern/testinfra/testrun/11821949216168733

Focused CUDA foreach parity, peak-memory, and large-tensor tests:
`buck2 --isolation-dir codex-muon-bench test fbcode//mode/opt fbcode//caffe2/test:test_optim_cuda -- --regex 'test_(foreach_matches_forloop|peak_memory_foreach|foreach_large_tensor).*Muon'`
Result: 4 passed, 0 failed.
TestInfra: https://www.internalfb.com/intern/testinfra/testrun/8162774698303264

CUDA automatic-default dispatch test:
`buck2 --isolation-dir codex-muon-bench test fbcode//mode/opt fbcode//caffe2/test:test_optim_cuda -- --regex 'test_defaults_changed_to_foreach.*Muon'`
Result: 2 passed, 0 failed.
TestInfra: https://www.internalfb.com/intern/testinfra/testrun/6755399814723046

Strengthened CUDA large-tensor test:
`buck2 --isolation-dir codex-muon-bench test fbcode//mode/opt fbcode//caffe2/test:test_optim_cuda -- --regex 'test_foreach_large_tensor.*Muon'`
Result: 2 passed, 0 failed.
TestInfra: https://www.internalfb.com/intern/testinfra/testrun/8725724653117781

Lint: `arc lint` passed for all modified fbcode/xplat files.

H100 steady-state CUDA microbenchmark:
`buck2 --isolation-dir codex-muon-bench run fbcode//mode/opt fbcode//caffe2/benchmarks/operator_benchmark:muon_optimizer_benchmark -- --tag-filter long --device cuda --warmup-iterations 10 --iterations 20 --num-runs 10`

Positive percentages mean foreach is faster than the for-loop implementation; each result is the mean of 10 runs.
- 1 rectangular parameter: foreach was 0.38%-2.03% slower.
- 2 parameters: foreach was 2.99%-6.64% faster.
- 3 parameters: foreach was 4.42%-12.71% faster.
- 4 parameters: foreach was 5.84%-7.87% faster.
- 16 parameters: foreach was 9.61%-13.40% faster.

The benchmark covers rectangular `512x1024`, `512x1536`, and `512x2048` parameters plus square `1024x1024` parameters for multi-parameter cases.

Differential Revision: D105621588

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
I'm of the opinion foreach none should go to the new caller, surprised anyone is using the incomplete Muon optimizer torch code in production

File: torch/optim/_muon.py [hunk 1]
```diff
@@ -362,10 +440,9 @@ def muon(
 
     See :class:`~torch.optim.Muon` for details.
     """
-    if foreach is not None and foreach:
-        raise RuntimeError("Foreach is not supported for Muon yet")
-
-    func = _single_tensor_muon
+    # foreach=None preserves the historical single-tensor default; only an
+    # explicit True opts into the multi-tensor (_foreach_*) path.
+    func = _multi_tensor_muon if foreach else _single_tensor_muon
```
Comment (janeyx99) [inline torch/optim/_muon.py:464]: Do you have benchmarks showing the performance on CUDA? Do we expect foreach to always be faster? If so, why not change the default in this PR?

File: test/test_optim.py [hunk 2]
```diff
@@ -2534,5 +2539,63 @@ def fwd_bwd(optim, mod, i):
...
+        opt2 = Muon([p2])
+        opt2.load_state_dict(sd)
+
+        # __setstate__ must backfill foreach to None, preserving the
+        # historical single-tensor default for older checkpoints.
+        self.assertEqual(opt2.param_groups[0]["foreach"], None)
+        # And a subsequent step() must succeed on the single-tensor path.
+        opt2.step()
+
+    def test_foreach_matches_forloop_bit_parity_rectangular(self):
```
Comment (janeyx99) [inline test/test_optim.py]: Please use OptimInfo tests similar to other tests in this file to be able to test all Muon configs vs just one

File: test/test_optim.py [hunk 3]
```diff
@@ -2534,5 +2539,63 @@ def fwd_bwd(optim, mod, i):
 instantiate_device_type_tests(TestSWAUtils, globals(), allow_mps=True)
 
 
+class TestMuonForeach(TestCase):
+    """Tests for the bit-parity and state-dict BC contracts that the
+    multi-tensor (foreach) path advertises but that the generic OptimizerInfo
+    matrix does not test strictly enough."""
+
+    def test_state_dict_round_trip_backfills_foreach(self):
```
Comment (janeyx99) [inline test/test_optim.py]: I suspect this test is redundant with another similar test already in this file, please co-opt that test instead of adding one offs.

**Comment by tangyuhao** [comment]
synced to master, and added benchmark and remove test cases that is duplicate

**Comment by tangyuhao** [comment]
@janeyx99 do you know how should i locally verify those failing tests? in fbcode i cannot get the same failure and looks like it is environment specific.

File: test/test_optim.py [hunk 4]
```diff
@@ -850,14 +850,19 @@ def _test_derived_optimizers(
...
-                # ref: https://github.com/pytorch/pytorch/issues/100701
-                empty_param = torch.empty(
-                    (), device=device, dtype=dtype, requires_grad=True
-                )
-                empty_param.grad = torch.rand_like(empty_param)
-                params = list(model.parameters()) + [empty_param]
+                if optim_cls.__name__ == "Muon":
+                    # Muon only accepts 2D parameters, so filter out the 1D
+                    # biases and the empty/0-D scalar.
+                    params = [p for p in model.parameters() if p.ndim == 2]
```
Comment (janeyx99) [inline test/test_optim.py:869]:
```suggestion
                    # Muon only accepts 2D parameters, so filter out the 1D
                    # biases
                    params = [p for p in model.parameters() if p.ndim == 2]
```
there should be no empty tensor then

File: test/test_optim.py [hunk 5]
```diff
@@ -1090,6 +1095,8 @@ def test_peak_memory_foreach(self, device, dtype, optim_info):
                 if optim_cls.__name__ == "Adafactor" and kwargs.get("maximize", False):
                     # When maximize is True, Adafactor also tracks device_grad
                     nintermediates = 3
+            elif optim_cls.__name__ == "Muon":
+                nintermediates = 2
```
Comment (janeyx99) [inline test/test_optim.py:1172]: why?

File: test/test_optim.py [hunk 6]
```diff
@@ -1614,7 +1624,10 @@ def _get_model_and_input_tensor(device, dtype, optim_cls):
         for optim_input in all_optim_inputs:
             torch.manual_seed(1)
             model, input = _get_model_and_input_tensor(device, dtype, optim_cls)
-            optimizer = optim_cls(model.parameters(), **optim_input.kwargs)
+            params = model.parameters()
+            if optim_cls.__name__ == "Muon":
+                params = [{"params": [p]} for p in model.parameters()]
```
Comment (janeyx99) [inline test/test_optim.py]: why

File: test/test_optim.py [hunk 7]
```diff
@@ -1638,6 +1651,9 @@ def fwd_bwd(optim, mod, i):
                         del group[flag]
 
             optimizer.load_state_dict(old_state_dict)
+            if optim_cls.__name__ == "Muon":
+                for group in optimizer.param_groups:
+                    self.assertEqual(group["foreach"], None)
```
Comment (janeyx99) [inline test/test_optim.py]: why is this specialized?
i'd expect this to be tested already with some more central logic

File: test/test_optim.py [hunk 8]
```diff
@@ -2288,13 +2304,17 @@ def test_fused_cpu_matches_device(self, device, dtype, optim_info):
...
+                and o.optim_cls.__name__ not in ("Adafactor", "Muon")
+            )
         ],
         dtypes=[torch.float32],
     )
     def test_defaults_changed_to_foreach(self, device, dtype, optim_info):
         # Test that the default implementations for optimizers are changed to foreach
-        # except Adafactor, which defaults to the single tensor impl for memory efficiency.
+        # except Adafactor, which defaults to the single tensor impl for memory efficiency,
+        # and Muon, whose foreach path is explicit opt-in for BC.
```
Comment (janeyx99) [inline test/test_optim.py]: why explicit opt in

File: torch/testing/_internal/common_optimizers.py [hunk 9]
```diff
@@ -1933,8 +1933,10 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
         Muon,
         optim_inputs_func=optim_inputs_func_muon,
         optim_error_inputs_func=optim_error_inputs_func_muon,
-        supported_impls=(),
-        not_og_supported_flags=(),
+        supported_impls=("foreach",),
+        # `foreach` was not in the original Muon release; it must be exercised by
+        # the BC tests that compare new impls against the original (single-tensor).
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py]:
```suggestion
```
this is implicit in the variable name

File: torch/testing/_internal/common_optimizers.py [hunk 10]
```diff
@@ -1948,6 +1950,13 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
                 "CompiledOptimizerParityTests",
                 "test_correctness",
             ),
+            DecorateInfo(
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py]: why is this a code skip instead of being addressed in the test case itself like the other tests?

**Review by janeyx99 — COMMENTED** [review]
Left a preliminary round of reviews, I'd like to see your benchmark results and why we're not making this codepath the default.

It should not be BC to make it the default.
